### PR TITLE
:sparkles: Dataset parser

### PR DESCRIPTION
**Pull request by Axelvel**
Added functions to extract the data of an .evrp file and create an ECVRPInstance with these parameters.

## What has been done

A few utils methods have been added to manipulate our datasets:

- get_datasets() that fetches all files/datasets in a given directory (defaults to where our datasets are stored)
- parse_dataset() that parses a given .evrp file and create a dictionary with the extracted data
- compute_distance_matrix() that create a 2D list in which is stored the Euclidean distance between each node
- create_ecvrp() that creates an ECVRPInstance with the given parsed data

## What needs to be done

- [x] Decide what if we should translate the ID of the list to their position in the list/matrix/dict or keep it this way → Fixed by offsetting each city ID
- [x] Modify ECVRPInstance method get_towns() → returns a list starting with an index 0, even if not town has this ID → Fixed with the ID offset
- [x] Adapt the test_utils suite to reflect these changes

## Coverage
|Name  |  Stmts |  Miss | Cover | Missing
|-|-|-|-|-|
|TOTAL |    354  |    0 |  100%|

### CODE REVIEW

File: src/server/utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+def compute_distance_matrix(nodes: dict[int, tuple[int, int]]) -> list[list[float]]:
+    """ Compute the distance matrix of our nodes
+    """
+    distance_matrix = np.zeros(shape=(len(nodes), len(nodes))).tolist()
+
+    for node, (current_x, current_y) in nodes.items():
+        for next_node, (next_x, next_y) in nodes.items():
+            if node != next_node:
+                dist = np.sqrt(np.power(current_x-next_x, 2) + np.power(current_y-next_y, 2))
+                distance_matrix[node-1][next_node-1] = dist
```
Comment (galyfray) [inline src/server/utils/utils.py:142]: I think that either the nodes have to be offsetted when reading or the offset should be parametrized. This issue has already been discussed, but nothing was decided about what to do.

File: test/server/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+DATASET = 'E-n5-k4-s2.evrp'
+
+cities = {1: (145, 215), 2: (151, 264), 3: (159, 261), 4: (130, 254), 5: (200, 176)}
+
+parsed_data = {
+    'VEHICLES': 4,
+    'CAPACITY': 6000,
+    'ENERGY_CAPACITY': 99,
+    'ENERGY_CONSUMPTION': 1.0,
+    'NODES': {1: (145, 215), 2: (151, 264), 3: (159, 261), 4: (130, 254), 5: (200, 176)},
```
Comment (galyfray) [inline test/server/test_utils.py:41]: Nodes and cities are duplicated. 
A possible fix could be :
```suggestion
    'NODES': cities,
```

File: test/server/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    'DEMANDS': {1: 0, 2: 1100, 3: 700},
+    'STATIONS': {4, 5},
+    'TIME_WINDOWS': {
+        1: (0, float('inf')), 2: (0, float('inf')), 3: (0, float('inf')),
+        4: (0, float('inf')), 5: (0, float('inf'))},
+    'DEPOT': 1}
+
+distance_matrix = [
```
Comment (galyfray) [inline test/server/test_utils.py:49]: Not a big fan of this matrix. Pytest has a way to assert that numbers are close enough.

File: test/server/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    [48.08326112068523, 8.54400374531753, 0.0, 29.832867780352597, 94.37160589923221],
+    [41.78516483155236, 23.259406699226016, 29.832867780352597, 0.0, 104.80458005259122],
+    [67.42403132415029, 100.72239075796404, 94.37160589923221, 104.80458005259122, 0.0]
+]
+
+
+def test_get_datasets():
+    """ Test if it lists the datasets for a given path.
+    """
+
```
Comment (galyfray) [inline test/server/test_utils.py:61]: They should not be any blank lines betweens function's docstring and the function's code. (replacing pydocstyle once again.

File: test/server/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    """ Test if the EVRPInstance is created with the right parameters.
+    """
+
+    evrp = create_ecvrp(parsed_data)
+
+    assert evrp.get_ev_capacity() == 6000
+    assert evrp.get_demand(2) == 1100
+    assert evrp.get_ev_count() == 4
+    assert evrp.get_tw(2) == (0, float('inf'))
+    # assert evrp.get_towns() == [2, 3]  # TODO: Fix get_towns()
```
Comment (galyfray) [inline test/server/test_utils.py:83]: This should be indicated in the PR

**Review by galyfray — CHANGES_REQUESTED** [review]
An overall good request. I note that all of your code constantly use single quote, while the rest of the codebase use double quotes. This isn't much of a problem, just noting. 
The description of the PR needs to be improved also.  Please add more detail over what you have done, what you did not do. Include the possible problems or things that will have to be done. Please also provide a copy of the table provided by the make coverage command so that everyone can see how much the tests are effective quickly. Note that you will have to do slight adjustment to the table for it to be recognized as a Markdown table :).

**Comment by galyfray** [comment]
I have also noted that some issue raised by the codacy's code analysis can be fixed.

File: src/server/utils/utils.py [hunk 6]
```diff
@@ -113,33 +121,41 @@ def create_ecvrp(parameters: dict[str, any]) -> ECVRPInstance:
...
     distance_matrix = np.zeros(shape=(len(nodes), len(nodes))).tolist()
 
     for node, (current_x, current_y) in nodes.items():
         for next_node, (next_x, next_y) in nodes.items():
             if node != next_node:
                 dist = np.sqrt(np.power(current_x-next_x, 2) + np.power(current_y-next_y, 2))
-                distance_matrix[node-1][next_node-1] = dist
-                distance_matrix[next_node-1][node-1] = distance_matrix[node-1][next_node-1]
+                distance_matrix[node][next_node] = dist
+                distance_matrix[next_node][node] = distance_matrix[node][next_node]
```
Comment (galyfray) [inline src/server/utils/utils.py:159]: Since you iterate on the whole matrix you should not need this.

File: src/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+"""Python package"""
```
Comment (galyfray) [inline src/__init__.py:1]:
```suggestion
"""Python package."""
```

File: src/server/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+"""Python package"""
```
Comment (galyfray) [inline src/server/__init__.py:1]:
```suggestion
"""Python package."""
```

**Review by galyfray — CHANGES_REQUESTED** [review]
Great work ! The automatic analysis found some small problems which can be fixed (at least a missing type hint, but I didn't get through every note left by the tool). 
I also reviewed the check that mostly fail because of bad documentation, which has been fixed in another branch.

**Review by galyfray — APPROVED** [review]
Everything seems fine

**Review by galyfray — APPROVED** [review]

**Review by Nexowo — APPROVED** [review]
Seems to be good, didn't notice any obvious mistake

**Pull request merged**